src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+	 *
+	 * Note: Assumes passed in events are already sorted.
+	 *
+	 * @param events events with phase info
+	 * @returns
+	 */
+	accountEventsByAddress(events: PActionEvent[]): Map<string, PAccountEvent[]> {
+		return events.reduce((acc, cur) => {
+			if (
+				!(
+					(cur.storagePath as PalletKeyInfo).module == 'system' &&
+					(cur.storagePath as PalletKeyInfo).item == 'account'
+				)
+			) {
+				// filter out non system::account events
+				return acc;
+			}
+
+			const asAccount = this.toAccountEvent(cur);
+			const address = asAccount.address.toString();
+
+			if (!acc.has(address)) {
+				acc.set(address, [asAccount]);
+			} else {
+				acc.get(address)?.push(asAccount);
+			}
+
+			return acc;
+		}, new Map<string, PAccountEvent[]>());
+	}
+
+	/**
+	 * Convert an `PActionEvent` to a `PAccountEvent` by adding the decoded `accountInfo`
+	 * and `address`.
+	 *
+	 * Notes: will throw if event does not have system::account storagePath.
+	 *
+	 * @param event PActionEvent
+	 */
+	private toAccountEvent(event: PActionEvent): PAccountEvent {
+		const { storagePath } = event;
+		if (
+			!(
+				(storagePath as PalletKeyInfo).module == 'system' &&
+				(storagePath as PalletKeyInfo).item == 'account'
+			)
+		) {
+			throw new InternalServerError(
+				'Event did not have system::account path as expected'
+			);
+		}
+
+		// storage key = h(system) + h(account) + h(address) + address
+		//								^^^^^^^^^^^^^^^^^^^^storage item prefix
+		// Since this is a transparent key with the address at the end we can pull out
+		// the address from the key. Here we slice off the storage key + account hash.
+		const addressRaw = event.data.stringValues.key?.slice(96);
+		if (!(typeof addressRaw === 'string')) {
+			throw new InternalServerError(
+				'Expect encoded address in system::account event storage key'
+			);
+		}
+		const address = this.registry.createType('Address', `0x${addressRaw}`);
+
+		const accountInfoEncoded = event?.data?.stringValues?.result;
+		if (!(typeof accountInfoEncoded === 'string')) {
+			throw new InternalServerError(
+				'Expect accountInfoEncoded to always be a string in system::Account event'
+			);
+		}
+		const accountInfo = (this.registry.createType(
+			'AccountInfo',
+			`0x${accountInfoEncoded.slice(2)}` // Slice off the leading Option byte
+		) as unknown) as AccountInfo;
+
+		return { ...event, accountInfo, address };
+	}
+
+	/**
+	 * Derive `Operation`s based on consecutive Get/Put events to an addresses `accountInfo`
+	 * storage item.
+	 *
+	 * `Operation`s are created by checking for deltas in each field of `accountInfo` based
+	 * on the previous event vs the current event.
+	 *
+	 * @param accountEventsByAddress map of address => events of that addresses `accountInfo`
+	 * events.
+	 */
+	private deriveOperations(
+		accountEventsByAddress: Map<string, PAccountEvent[]>
+	): Operation[] {
+		const ops = [];
+		for (const events of accountEventsByAddress.values()) {
```

Comment:
You would know best, but just throwing it out there. Is a double for loop here absolutely necessary? (Assuming it is), but if it wasn't and there was a way to optimize it would it be worth it? As in is the current size of event's always going to be so small that it wont matter, or is the potential size of events large enough to cause issues scaling.

Reply:
Unless there is an error to the logic, any given event will only belong to one address. So we loop through each address, and for each address we loop through its events. Because an event can only belong to one address we will never see the same event twice. Thus the worst case complexity of these nested loops is equal to the number of events.